experiments/tf_trainer/common/model_trainer.py

```diff
@@ -41,32 +51,120 @@
+
+  def verify_key_types(keys):  # pylint: disable=missing-docstring
+    if keys is None:
+      return keys
+    if isinstance(keys, six.string_types):
+      return [keys]
+    if not isinstance(keys, (list, tuple)):
+      raise TypeError('keys should be either a string or a list of strings. '
+                      'Given: {}'.format(type(keys)))
+    for key in keys:
+      if not isinstance(key, six.string_types):
+        raise TypeError('All items in the given keys list should be a string. '
+                        'There exist an item with type: {}'.format(type(key)))
+    return keys
+
+  def get_keys(features):
+    if keys is None:
+      return features.keys()
+    return keys
+
+  def verify_keys_and_predictions(features, predictions):
+    if not isinstance(predictions, dict):
+      raise ValueError(
+          'Predictions should be a dict to be able to forward features. '
+          'Given: {}'.format(type(predictions)))
+    for key in get_keys(features):
+      if key not in features:
+        raise ValueError(
+            'keys should be exist in features. Key "{}" is not in features '
+            'dict. features dict has following keys: {}. Please check '
+            'arguments of forward_features.'.format(key, features.keys()))
+      if key in predictions:
+        raise ValueError(
+            'Cannot forward feature key ({}). Since it does exist in '
+            'predictions. Existing prediction keys: {}. Please check arguments '
+            'of forward_features.'.format(key, predictions.keys()))
+
+  keys = verify_key_types(keys)
+
+  def new_model_fn(features, labels, mode, config):  # pylint: disable=missing-docstring
+    spec = estimator.model_fn(features, labels, mode, config)
+    predictions = spec.predictions
+    if predictions is None:
+      return spec
+    verify_keys_and_predictions(features, predictions)
+    for key in get_keys(features):
+      feature = sparse_tensor_lib.convert_to_tensor_or_sparse_tensor(
+          features[key])
+      if sparse_default_values and (key in sparse_default_values):
+        if not isinstance(feature, sparse_tensor_lib.SparseTensor):
+          raise ValueError(
+              'Feature ({}) is expected to be a `SparseTensor`.'.format(key))
+        feature = sparse_ops.sparse_tensor_to_dense(
+            feature, default_value=sparse_default_values[key])
+      if not isinstance(feature, ops.Tensor):
+        raise ValueError(
+            'Feature ({}) should be a Tensor. Please use `keys` '
+            'argument of forward_features to filter unwanted features, or'
+            'add key to argument `sparse_default_values`.'
+            'Type of features[{}] is {}.'.format(key, key, type(feature)))
+      predictions[key] = feature
+    spec = spec._replace(predictions=predictions)
+    if spec.export_outputs: # CHANGES HERE
+      outputs = spec.export_outputs['predict'].outputs
+      outputs['comment_key'] = spec.predictions['comment_key']
```

Comment:
Where do the keys like 'predict' and 'comment_key' come from? Are they hardcoded for our purposes or some kind of general keys used by EstimatorSpec?

Reply:
Good catch. comment_key was definitely an error from my end.

for 'predict' and 'serving_default' these are general keys by estimatorSpec. Let me know if I can leave them as is